android/engine/src/main/java/org/smartregister/fhircore/engine/task/WorkflowCarePlanGenerator.kt

```diff
@@ -0,0 +1,329 @@
+    outcomeReferences: List<Reference>,
+    updateCarePlan: Boolean = true,
+  ) {
+    val carePlanActivityStatus: CarePlan.CarePlanActivityStatus
+    val carePlan: CarePlan
+    when (requestResource.fhirType()) {
+      "Task" -> {
+        updateRequestResourceStatus(requestResource as Task, requestResourceStatus)
+        if (updateCarePlan) {
+          carePlanActivityStatus = mapRequestResourceStatusToCarePlanStatus(requestResource)
+          carePlan =
+            if (requestResource.hasBasedOn()) {
+              fhirEngine.get(
+                ResourceType.CarePlan,
+                IdType(requestResource.basedOnFirstRep.referenceElement.value).idPart,
+              ) as CarePlan
+            } else {
+              return
+            }
+          updateCarePlanStatus(carePlan, requestResource, carePlanActivityStatus, outcomeReferences)
+        }
+      }
+      "ServiceRequest" -> TODO("Not supported yet")
+      "MedicationRequest" -> TODO("Not supported yet")
+      "SupplyRequest" -> TODO("Not supported yet")
+      "Procedure" -> TODO("Not supported yet")
+      "DiagnosticReport" -> TODO("Not supported yet")
+      "Communication" -> TODO("Not supported yet")
+      "CommunicationRequest" -> TODO("Not supported yet")
+      "RequestGroup" -> {}
+      else -> TODO("Not a valid request resource")
+    }
+  }
+
+  /** Update [Task] status */
+  suspend fun updateRequestResourceStatus(resource: Task, status: String) {
+    resource.status = Task.TaskStatus.valueOf(status)
+    resource.executionPeriod.end = Date.from(Instant.now())
+    fhirEngine.update(resource)
+  }
+
+  /** Map [Task] status to [CarePlan] status */
+  fun mapRequestResourceStatusToCarePlanStatus(
+    resource: Task,
+  ): CarePlan.CarePlanActivityStatus {
+    // Refer: http://hl7.org/fhir/R4/valueset-care-plan-activity-status.html for some mapping
+    // guidelines
+    return when (resource.status) {
+      Task.TaskStatus.ACCEPTED -> CarePlan.CarePlanActivityStatus.SCHEDULED
+      Task.TaskStatus.DRAFT -> CarePlan.CarePlanActivityStatus.NOTSTARTED
+      Task.TaskStatus.REQUESTED -> CarePlan.CarePlanActivityStatus.NOTSTARTED
+      Task.TaskStatus.RECEIVED -> CarePlan.CarePlanActivityStatus.NOTSTARTED
+      Task.TaskStatus.REJECTED -> CarePlan.CarePlanActivityStatus.STOPPED
+      Task.TaskStatus.READY -> CarePlan.CarePlanActivityStatus.NOTSTARTED
+      Task.TaskStatus.CANCELLED -> CarePlan.CarePlanActivityStatus.CANCELLED
+      Task.TaskStatus.INPROGRESS -> CarePlan.CarePlanActivityStatus.INPROGRESS
+      Task.TaskStatus.ONHOLD -> CarePlan.CarePlanActivityStatus.ONHOLD
+      Task.TaskStatus.FAILED -> CarePlan.CarePlanActivityStatus.STOPPED
+      Task.TaskStatus.COMPLETED -> CarePlan.CarePlanActivityStatus.COMPLETED
+      Task.TaskStatus.ENTEREDINERROR -> CarePlan.CarePlanActivityStatus.ENTEREDINERROR
+      Task.TaskStatus.NULL -> CarePlan.CarePlanActivityStatus.NULL
+      else -> CarePlan.CarePlanActivityStatus.NULL
+    }
+  }
```

Comment:
I might be missing some but why are we doing this?  A CarePlan can have multiple tasks so it's status is not mapped 1 to 1 to a task.

Reply:
https://github.com/opensrp/fhircore/blob/e7860d69e8766b4caa33c0e8b82b0bcb03001d51/android/engine/src/main/java/org/smartregister/fhircore/engine/task/WorkflowCarePlanGenerator.kt#L151-L165

creating new activity on each task

Reply:
ah ok got it, so this is setting the status of each task-related activity in the careplan, not the status of the activity itself